test/unit/context/test-node-values.js

```diff
@@ -0,0 +1,878 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ContextNode} from '../../../src/context/node';
+import {Values} from '../../../src/context/values';
+import {contextProp} from '../../../src/context/prop';
+
+const NonRecursive = contextProp('NonRecursive');
+
+const Recursive = contextProp('Recursive', {needsParent: true});
+
+const Concat = contextProp('Concat', {
+  needsParent: true,
+  compute: (contextNode, inputs, parentValue) =>
+    `${parentValue}${inputs.length > 1 ? `(${inputs.join('|')})` : inputs[0]}`,
```

Comment:
The property value is a function `const usedValue = compute(input)`. This is analogous to the CSS's "specified" vs "used" values concept. For instance, you can set `font-size: 80%` and it'll be computed using inherited value and 80% expression. It's similar here as well. If I were to set `node.set(renderable, true)` - that's an input. The actual value can still compute to `used.rendrable === false` because one of the parents still has `false` value. This is covered in more detail in the design doc as well, if that helps.